### PR TITLE
READY: Make ByIndex.index type of SInt

### LINKED ISSUE #99: ByIndex problems
- [x] index should not be Scala Int. `Value[SInt.type]` expected
- [ ] default value may me useful

relates to #87, #88, #98

### PR DESCRIPTION
**Pull request by greenhat**
Relates to #99

### CODE REVIEW

**Comment by greenhat** [comment]
Ready for review. I'll continue with #99 (`ByIndex` default value) is separate PR.

**Comment by catena2w** [comment]
There is special "ready for merge" label

File: src/main/scala/sigmastate/serialization/transformers/ByIndexSerializer.scala [hunk 1]
```diff
@@ -15,12 +15,13 @@ object ByIndexSerializer extends ValueSerializer[ByIndex[SType]] {
   override val opCode: OpCode = OpCodes.ByIndexCode
 
   override def parseBody(bytes: Array[Byte], pos: Position): (ByIndex[SType], Consumed) = {
-    val (input, consumed) = ValueSerializer.deserialize(bytes, pos)
-    val index = Ints.fromByteArray(bytes.slice(pos + consumed, pos + consumed + intLength))
```
Comment (catena2w) [inline src/main/scala/sigmastate/serialization/transformers/ByIndexSerializer.scala]: Don't forget to remove unused imports like `com.google.common.primitives.Ints`

Comment (greenhat) [reply src/main/scala/sigmastate/serialization/transformers/ByIndexSerializer.scala]: @catena2w thanks! Fixed.

**Comment by greenhat** [comment]
@catena2w I love labels but they are disabled for me. Along with milestones, etc. I need a write access to the repo for this.

**Comment by greenhat** [comment]
Rebased and resolved conflicts.

**Comment by greenhat** [comment]
@catena2w Rebased and fixed.

File: src/main/scala/sigmastate/utxo/transformers.scala [hunk 2]
```diff
@@ -247,23 +247,22 @@ object Fold {
...
+  override def reconstruct(cs: immutable.Seq[Any]): ByIndex[V] = cs match {
+    case Seq(input: Value[SCollection[V]]@unchecked, index: Value[SInt.type]@unchecked, _) =>
+      ByIndex[V](input, index)
+    case _ => illegalArgs("ByIndex", "(Value[SCollection[V]], index: Value[SInt.type])(tpe: V)", cs)
   }
 
-  override def function(input: EvaluatedValue[SCollection[V]]) =
-    input.items.apply(index)
+  override def function(input: EvaluatedValue[SCollection[V]]): Value[V] =
+    input.items.apply(index.asInstanceOf[EvaluatedValue[SInt.type]].value.toInt)
```
Comment (kushti) [inline src/main/scala/sigmastate/utxo/transformers.scala]: I see no any guarantee why index would be evaluated at the moment of function() being calling. I guess transformationReady() should also be redefined. And tests for index evaluation should be done.

Comment (greenhat) [reply src/main/scala/sigmastate/utxo/transformers.scala]: Oops, I missed that big time. Thanks. On it.

Comment (greenhat) [reply src/main/scala/sigmastate/utxo/transformers.scala]: @kushti Fixed.

**Pull request merged**